modules/siddhi-core/src/main/java/org/wso2/siddhi/core/util/parser/StoreQueryParser.java

```diff
@@ -77,59 +91,129 @@
+                        within = aggregationInputStore.getWithin();
+                        per = aggregationInputStore.getPer();
+                    } else if (aggregationInputStore.getPer() != null || aggregationInputStore.getWithin() != null) {
+                        throw new StoreQueryCreationException(
+                                inputStore.getStoreId() + " should either have both 'within' " +
+                                        "and 'per' defined or none.");
+                    }
+                    if (((AggregationInputStore) inputStore).getOnCondition() != null) {
+                        onCondition = ((AggregationInputStore) inputStore).getOnCondition();
+                    }
+                } else if (inputStore instanceof ConditionInputStore) {
+                    if (((ConditionInputStore) inputStore).getOnCondition() != null) {
+                        onCondition = ((ConditionInputStore) inputStore).getOnCondition();
+                    }
                 }
-            } else if (inputStore instanceof ConditionInputStore) {
-                if (((ConditionInputStore) inputStore).getOnCondition() != null) {
-                    onCondition = ((ConditionInputStore) inputStore).getOnCondition();
+                List<VariableExpressionExecutor> variableExpressionExecutors = new ArrayList<>();
+                table = tableMap.get(inputStore.getStoreId());
+                if (table != null) {
+                    return constructStoreQueryRuntime(table, storeQuery, siddhiAppContext, tableMap, windowMap,
+                            queryName,
+                            metaPosition, onCondition, metaStreamEvent, variableExpressionExecutors, lockWrapper);
+                } else {
+                    AggregationRuntime aggregation = aggregationMap.get(inputStore.getStoreId());
+                    if (aggregation != null) {
+                        return constructStoreQueryRuntime(aggregation, storeQuery, siddhiAppContext, tableMap, queryName,
+                                within, per, onCondition, metaStreamEvent, variableExpressionExecutors);
+                    } else {
+                        Window window = windowMap.get(inputStore.getStoreId());
+                        if (window != null) {
+                            return constructStoreQueryRuntime(window, storeQuery, siddhiAppContext, tableMap, queryName,
+                                    metaPosition, onCondition, metaStreamEvent, variableExpressionExecutors);
+                        } else {
+                            throw new StoreQueryCreationException(
+                                    inputStore.getStoreId() + " is neither a table, aggregation or window");
+                        }
+                    }
                 }
+
+            } finally {
+                SnapshotService.getSkipSnapshotableThreadLocal().set(null);
             }
+        } else if (storeQuery.isDeleteQuery()) {
```

Comment:
introduce an enum and use switch statement (to improve performance)

Reply:
Fixed with e206a5089d27e5abbd49123fe002f2426eba9cd8